File: backend/src/file_ops.rs
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::env;
/**
 * File Operations Module
 *
 * Tauri commands for file I/O operations:
 * - Open and read Cell files (CSV, TSV, JSON)
 * - Save Cell files
 * - Get file identifiers for config matching
 * - Calculate partial content hashes
 *
 * Note: In Tauri 2.0, file dialogs are handled by the tauri-plugin-dialog
 * from the frontend instead of custom Rust commands.
 */
use std::fs;
use std::io::Read;
use std::path::{Path, PathBuf};

/// Common path validation checks for preventing path traversal attacks
/// Returns a PathBuf if basic validation passes, or an error if the path is suspicious
fn validate_path_basics(path: &str) -> Result<PathBuf, String> {
    // Check for suspicious patterns that indicate path traversal
    if path.contains("..") {
        return Err("Path traversal detected: '..' is not allowed in file paths".to_string());
    }

    // Check for null bytes (can be used to bypass checks)
    if path.contains('\0') {
        return Err("Invalid path: null bytes are not allowed".to_string());
    }

    // Create path and check if it's absolute
    let path_buf = PathBuf::from(path);

    // For file operations, we require absolute paths
    if !path_buf.is_absolute() {
        return Err("Only absolute file paths are allowed".to_string());
    }

    Ok(path_buf)
}

/// Verify a canonicalized path doesn't contain path traversal patterns
fn verify_canonical_path(canonical: &Path) -> Result<(), String> {
    if canonical.to_string_lossy().contains("..") {
        return Err("Path traversal detected after canonicalization".to_string());
    }
    Ok(())
}

/// Validates a file path to prevent path traversal attacks
/// Returns the canonicalized path if valid, or an error if the path is suspicious
fn validate_file_path(path: &str) -> Result<PathBuf, String> {
    let path_buf = validate_path_basics(path)?;

    // Canonicalize the path to resolve any symbolic links and get the real path
    let canonical = path_buf
        .canonicalize()
        .map_err(|e| format!("Failed to resolve path: {}", e))?;

    verify_canonical_path(&canonical)?;

    Ok(canonical)
}

/// Validates a file path for writing (file may not exist yet)
/// Uses parent directory validation to prevent path traversal
fn validate_file_path_for_write(path: &str) -> Result<PathBuf, String> {
    let path_buf = validate_path_basics(path)?;

    // Get the parent directory and validate it exists
    let parent = path_buf
        .parent()
        .ok_or("Invalid path: no parent directory")?;

    // Canonicalize the parent directory to ensure it exists and is valid
    let canonical_parent = parent
        .canonicalize()
        .map_err(|e| format!("Failed to resolve parent directory: {}", e))?;

    verify_canonical_path(&canonical_parent)?;

    // Get the filename and construct the final path
    let filename = path_buf.file_name().ok_or("Invalid path: no filename")?;

    Ok(canonical_parent.join(filename))
}

/// Read a Cell file (CSV, TSV, or JSON) from disk and return its contents as a string
#[tauri::command]
pub fn open_cell_file(path: String) -> Result<String, String> {
    // Validate path to prevent path traversal attacks
    let safe_path = validate_file_path(&path)?;

    fs::read_to_string(&safe_path).map_err(|e| format!("Failed to read file: {}", e))
}

/// Write Cell file content (CSV, TSV, or JSON) to a file on disk
#[tauri::command]
pub fn save_cell_file(path: String, content: String) -> Result<(), String> {
    // Validate path to prevent path traversal attacks
    let safe_path = validate_file_path_for_write(&path)?;

    fs::write(&safe_path, content).map_err(|e| format!("Failed to write file: {}", e))
}

/// Create a temporary Cell file and return its path
/// The file will be created in the system temp directory with a unique name
#[tauri::command]
pub fn create_temp_file() -> Result<String, String> {
    let temp_dir = env::temp_dir();

    // Create a unique temp file name with timestamp and nanoseconds for uniqueness
    // Using both seconds and nanoseconds prevents collisions when multiple files
    // are created in rapid succession
    let duration = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|e| format!("System time error: {}", e))?;

    let timestamp = duration.as_secs();
    let nanos = duration.subsec_nanos();

    let temp_file_name = format!("seria_temp_{}_{}.csv", timestamp, nanos);
    let temp_file_path = temp_dir.join(temp_file_name);

    // Create the file with empty CSV content (just headers)
    let initial_content = "Column1,Column2,Column3\n";
    fs::write(&temp_file_path, initial_content)
        .map_err(|e| format!("Failed to create temp file: {}", e))?;

    // Return the path as a string
    temp_file_path
        .to_str()
        .ok_or("Failed to convert temp file path to string".to_string())
        .map(|s| s.to_string())
}

/// File identifiers used for config matching
#[derive(Serialize, Deserialize)]
pub struct FileIdentifiers {
    pub absolute_path: String,
    pub filename: String,
    pub parent_dir: String,
    pub file_size: u64,
    pub content_hash_partial: Option<String>,
    pub os_file_id: Option<String>,
}

/// Get all identifiers for a file
/// These identifiers are used for matching file configs across renames and moves
#[tauri::command]
pub fn get_file_identifiers(path: String) -> Result<FileIdentifiers, String> {
    // Validate path to prevent path traversal attacks
    let safe_path = validate_file_path(&path)?;
    let path_obj = safe_path.as_path();

    // Get filename
    let filename = path_obj
        .file_name()
        .and_then(|n| n.to_str())
        .ok_or("Invalid filename")?
        .to_string();

    // Get parent directory name (not full path, just the immediate parent)
    let parent_dir = path_obj
        .parent()
        .and_then(|p| p.file_name())
        .and_then(|n| n.to_str())
        .unwrap_or("")
        .to_string();

    // Get file size
    let metadata =
        fs::metadata(&safe_path).map_err(|e| format!("Failed to get file metadata: {}", e))?;
    let file_size = metadata.len();

    // Get OS file ID (inode on Unix/Mac, File ID on Windows)
    let os_file_id = get_os_file_id(safe_path.to_str().unwrap_or(""));

    // Calculate partial content hash if file is large enough
    let content_hash_partial = if file_size > 100_000 {
        Some(calculate_content_hash(safe_path.to_str().unwrap_or(""))?)
    } else {
        None
    };

    Ok(FileIdentifiers {
        absolute_path: safe_path.to_string_lossy().to_string(),
        filename,
        parent_dir,
        file_size,
        content_hash_partial,
        os_file_id,
    })
}

/// Calculate partial content hash (first 1MB or entire file if smaller)
/// This helps identify files even after they're moved or renamed
fn calculate_content_hash(path: &str) -> Result<String, String> {
    let mut file =
        fs::File::open(path).map_err(|e| format!("Failed to open file for hashing: {}", e))?;

    let mut hasher = Sha256::new();
    let mut buffer = vec![0; 1024 * 1024]; // 1MB buffer

    // Read first 1MB (or entire file if smaller)
    let bytes_read = file
        .read(&mut buffer)
        .map_err(|e| format!("Failed to read file for hashing: {}", e))?;

    hasher.update(&buffer[..bytes_read]);
    let result = hasher.finalize();

    Ok(format!("{:x}", result))
}

/// Get OS-specific file ID (inode on Unix/Mac, File ID on Windows)
#[cfg(unix)]
fn get_os_file_id(path: &str) -> Option<String> {
    use std::os::unix::fs::MetadataExt;

    fs::metadata(path)
        .ok()
        .map(|m| format!("inode-{}", m.ino()))
}

#[cfg(windows)]
fn get_os_file_id(path: &str) -> Option<String> {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};

    // Note: The file_index() method is unstable in Rust, so we use a hash of the path
    // as a stable alternative. This provides a consistent identifier for the file.
    // In production, you might want to use the Win32 API to get the actual file index.
    let mut hasher = DefaultHasher::new();
    path.hash(&mut hasher);
    Some(format!("fileid-{}", hasher.finish()))
}

#[cfg(not(any(unix, windows)))]
fn get_os_file_id(_path: &str) -> Option<String> {
    // Unsupported platform
    None
}

File: apps/desktop/backend/src/file_ops.ts
/**
 * File operations for Cell files (CSV, TSV, JSON).
 *
 * Ported from backend/src/file_ops.rs. Shape of each command preserved so
 * the renderer's call sites can move to typed RPC with no behavioral
 * change.
 */

import { createHash } from "node:crypto";
import { promises as fs, readSync, openSync, closeSync, statSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, isAbsolute, dirname, basename, resolve } from "node:path";
import type { FileIdentifiers } from "@shared/rpc";

/**
 * Monotonic counter that, combined with a timestamp + pid, makes rapid
 * successive createTempFile calls collision-free.
 */
let tempFileCounter = 0;

function validatePathBasics(path: string): void {
	if (path.includes("..")) {
		throw new Error("Path traversal detected: '..' is not allowed in file paths");
	}
	if (path.includes("\0")) {
		throw new Error("Invalid path: null bytes are not allowed");
	}
	if (!isAbsolute(path)) {
		throw new Error("Only absolute file paths are allowed");
	}
}

async function validateFilePath(path: string): Promise<string> {
	validatePathBasics(path);
	const canonical = await fs.realpath(path);
	if (canonical.includes("..")) {
		throw new Error("Path traversal detected after canonicalization");
	}
	return canonical;
}

async function validateFilePathForWrite(path: string): Promise<string> {
	validatePathBasics(path);
	const parent = dirname(path);
	const canonicalParent = await fs.realpath(parent);
	if (canonicalParent.includes("..")) {
		throw new Error("Path traversal detected after canonicalization");
	}
	const filename = basename(path);
	if (!filename) throw new Error("Invalid path: no filename");
	return join(canonicalParent, filename);
}

export async function openCellFile({ path }: { path: string }): Promise<string> {
	const safe = await validateFilePath(path);
	return fs.readFile(safe, "utf-8");
}

export async function saveCellFile({
	path,
	content,
}: {
	path: string;
	content: string;
}): Promise<void> {
	const safe = await validateFilePathForWrite(path);
	await fs.writeFile(safe, content, "utf-8");
}

export async function createTempFile(): Promise<string> {
	const now = process.hrtime.bigint();
	// hrtime is nanoseconds since an arbitrary epoch; split into seconds/nanos
	// so the filename matches the Rust version's (secs, nanos, pid, counter)
	// shape. The exact epoch doesn't matter — uniqueness does.
	const ns = Number(now % 1_000_000_000n);
	const s = Number(now / 1_000_000_000n);
	const counter = tempFileCounter++;
	const name = `seria_temp_${s}_${ns}_${process.pid}_${counter}.csv`;
	const fullPath = join(tmpdir(), name);
	await fs.writeFile(fullPath, "Column1,Column2,Column3\n", "utf-8");
	return fullPath;
}

function calculateContentHashPartial(path: string): string {
	// First 1MB (or entire file if smaller), SHA-256'd. Matches the Rust
	// behavior of `read(&mut buffer)` with a 1MB buffer.
	const fd = openSync(path, "r");
	try {
		const buf = Buffer.alloc(1024 * 1024);
		const n = readSync(fd, buf, 0, buf.length, 0);
		return createHash("sha256").update(buf.subarray(0, n)).digest("hex");
	} finally {
		closeSync(fd);
	}
}

function getOsFileId(path: string): string | null {
	// Rust: inode on Unix, hash(path) on Windows. TS equivalent uses fs.stat's
	// ino on non-Windows (0 when unsupported) and a SHA256-of-path fallback on
	// Windows. Used to match file configs across renames/moves.
	if (process.platform === "win32") {
		const hash = createHash("sha256").update(path).digest("hex").slice(0, 16);
		return `fileid-${hash}`;
	}
	try {
		const s = statSync(path);
		if (s.ino === 0) return null;
		return `inode-${s.ino}`;
	} catch {
		return null;
	}
}

export async function getFileIdentifiers({
	path,
}: {
	path: string;
}): Promise<FileIdentifiers> {
	const safe = await validateFilePath(path);
	const filename = basename(safe);
	// Parent dir *name* only (not full path), matching the Rust semantics used
	// by file-config matching in the renderer.
	const parentDir = basename(dirname(safe));
	const meta = await fs.stat(safe);
	const fileSize = meta.size;
	const osFileId = getOsFileId(safe);
	const contentHashPartial =
		fileSize > 100_000 ? calculateContentHashPartial(safe) : null;
	return {
		absolutePath: resolve(safe),
		filename,
		parentDir,
		fileSize,
		contentHashPartial,
		osFileId,
	};
}
